### ISSUE TITLE
Open Question: feasibility of datastore using dynamodb

### ISSUE THREAD

**Issue by imacks** [authzed/spicedb#1083]
I'm looking for your thoughts on implementating a database backend for authzed in dynamodb. Zenzibar paper did specifically say that it is using spanner, but it reads like it is really zookie that plays the important role of guaranteeing causal acl. I don't have any concrete code right now, but my feeling is that it may be possible to design the db with mostly eventual consistency, with small amount of strong consistency parts. The caveat would be that eval node answers may be different from each other, abit only for a short period after write. This should be alright in real life use cases. Most IAM from big cloud providers are not strongly consistent.

Is there any technical difficulty I have overlooked?

**Comment by josephschorr** [comment]
@imacks The biggest open question is handling of zedtokens and "full consistency": if a zedtoken is provided, or full consistency is required, SpiceDB is required to honor that request by the client. Thus, there must either be a way in DynamoDB to request the fully consistent updates that have been applied *or* SpiceDB will have to wait until it can guarantee that the updates have been applied.

So long as this is doable (or it is acceptable to wait), I would imagine implementation should be fairly straightforward

**Comment by imacks** [comment]
Where strong consistency is required, we can use [conditional writes](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/WorkingWithItems.html#WorkingWithItems.ConditionalUpdate) and [strongly consistent reads](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ReadConsistency.html). Having said that, the biggest advantage I hope from DynamoDB backend is better latency as things scale up, which suggests an eventual consistency model.

**Comment by josephschorr** [comment]
Yeah, strong consistency will only be required when requested by the client. If the ZedToken's requirements are already met by the eventual consistency at the time of the call, or if `minimize_latency` is requested (which is the default), then eventual consistency is fine

**Comment by imacks** [comment]
Tks for the tip. I will try my hands on making a POC and then circle back. Cheers!

**Comment by josephschorr** [comment]
@imacks any progress? Quite curious :)

**Comment by josephschorr** [comment]
@imacks Ping!

**Comment by josephschorr** [comment]
@imacks Sounds great! Let us know if we can help in any way

**Comment by josephschorr** [comment]
@imacks Would you be willing to share your progress? I'm quite curious :)

**Comment by jzelinskie** [comment]
I just wanted to follow up on this thread as we've seen requests for DynamoDB usage in multi-region SpiceDB deployments.

DynamoDB can be used to implement a single-region deployment, but in a multi-region DynamoDB configuration, writes across regions are replicated in an eventually consistent, asynchronous fashion. This means requests with a `full_consistency` ZedToken in this configuration would produce incorrect results.

**Comment by josephschorr** [comment]
@imacks Checking in again to see if you have anything you'd mind sharing :)

**Comment by gsingh-ds** [comment]
I have quite a bit of experience with Dynamodb and Go. Also we have written a go lib around making working with Dynamo much easier. We were thinking to make it open source. @imacks happy to help with anything you need.

**Comment by benny-yamagata** [comment]
Pinging here to see if there have been any changes since earlier this year?  I have thought about taking a crack at this myself but dont want to pull the trigger if someone else has it mostly done.

**Comment by josephschorr** [comment]
@benny-yamagata As far as we know, no updates. Feel free to give it a shot and we'd be happy to work with you on it

**Comment by benny-yamagata** [comment]
> @benny-yamagata As far as we know, no updates. Feel free to give it a shot and we'd be happy to work with you on it

Highly considering it 😉

**Comment by josephschorr** [comment]
@benny-yamagata did you ever get a chance to try?

**Comment by benny-yamagata** [comment]
> [@benny-yamagata](https://github.com/benny-yamagata) did you ever get a chance to try?

@josephschorr I didnt, I have moved roles since the last time I posted and am not as close to the code anymore.  However, the discussion at my org has come up again around this and there is renewed interest in spice on dynamo.  

Even a paired down version for single region deployments.

**Comment by miparnisari** [comment]
FWIW, as of December 2024 there is a new feature (in preview) that allows for strongly consistent reads across regions: https://aws.amazon.com/about-aws/whats-new/2024/12/amazon-dynamodb-global-tables-previews-multi-region-strong-consistency/

**Comment by benny-yamagata** [comment]
@josephschorr , our use case not only does not require strongly consistent reads **across regions**, it prevents cross region communication entirely. 

Given this, is there a way to do a paired down version of spicedb for single region deployments? That way people with this use case can start utilizing spiceDB? Then we can focus on a longer term multi region solution? 
Just trying to figure out how to slice this up in a way where we can make progress.

**Comment by josephschorr** [comment]
@benny-yamagata feel free to start a prototype implementation; its fine to document it as not supporting multi-region

**Comment by Meyazhagan** [comment]
I’ve started the prototype implementation and will create a PR. I’m assuming no one else is currently working on this.

**Comment by benny-yamagata** [comment]
@Meyazhagan I see that the PR closed, were you not able to implement this?

**Comment by Meyazhagan** [comment]
Hi @benny-yamagata I didn’t get a chance to finish it earlier since I still needed to add tests and clean up some of the code. I have some time now and will reopen the PR later this week for review.
